File: programs/fluid-solana-programs/__tests__/vaults/combination.test.ts
import { expect } from "chai";
import { BN } from "@coral-xyz/anchor";
import { Keypair } from "@solana/web3.js";

import { mint as MintInfo } from "../../ts-sdk/mint";
import { VaultBaseSetup } from "../../test-utils/typescript/vaults/setup";

const DEFAULT_ORACLE_PRICE = new BN(10).pow(new BN(8));

async function createCheckedPosition(
  vaultId: number,
  collateral: BN,
  debt: BN,
  user: Keypair,
  setup: VaultBaseSetup,
  positionId?: number
) {
  if (!positionId) {
    positionId = await setup.getNextPositionId(vaultId);
    await setup.initPosition(vaultId, user);
  }

  const userSupplyTokenBalanceBefore = await setup.balanceOf(
    user.publicKey,
    MintInfo.getMint(setup.getVaultSupplyToken(vaultId))
  );

  const userBorrowTokenBalanceBefore = await setup.balanceOf(
    user.publicKey,
    MintInfo.getMint(setup.getVaultBorrowToken(vaultId))
  );

  await setup.operateVault({
    vaultId,
    positionId,
    user: user,
    positionOwner: user,
    collateralAmount: collateral,
    debtAmount: debt,
    recipient: user,
  });

  const userSupplyTokenBalanceAfter = await setup.balanceOf(
    user.publicKey,
    MintInfo.getMint(setup.getVaultSupplyToken(vaultId))
  );

  const userBorrowTokenBalanceAfter = await setup.balanceOf(
    user.publicKey,
    MintInfo.getMint(setup.getVaultBorrowToken(vaultId))
  );

  // round up the collateral during deposit
  if (collateral.gt(new BN(0))) {
    collateral = collateral.add(new BN(1));
  }

  // round down the debt during payback
  if (debt.lt(new BN(0))) {
    debt = debt.sub(new BN(1));
  }

  // prettier-ignore
  {
    expect(userSupplyTokenBalanceBefore.sub(userSupplyTokenBalanceAfter).toString()).to.be.eq(collateral.toString());
    expect(userBorrowTokenBalanceAfter.sub(userBorrowTokenBalanceBefore).toString()).to.be.eq(debt.toString());
  }

  return positionId;
}

describe("vaults - combination operations", () => {
  let setup: VaultBaseSetup;

  beforeEach(async () => {
    process.env.TEST_MODE_JEST = "true";

    setup = new VaultBaseSetup();
    await setup.setup();

    // set oracle price to be 1e18 for USDC/USDT
    setup.setOraclePrice(DEFAULT_ORACLE_PRICE, true);
  });

  afterEach(async () => {
    setup.logComputeBudget();
  });

  // prettier-ignore
  async function _assertState(
    vaultId: number,
    tick: number,
    tickDebt: BN,
    topmostTick: number,
    totalCollateral: BN,
    totalDebt: BN
  ) {
    const vaultState = await setup.vaultModule.readVaultState({vaultId});
    const decimalScaleFactorBorrowToken = setup.getDecimalScaleFactor(setup.getVaultBorrowToken(vaultId));
    const decimalScaleFactorSupplyToken = setup.getDecimalScaleFactor(setup.getVaultSupplyToken(vaultId));

    const topTick = vaultState.topmostTick.toString();
    const vaultDebt = new BN(vaultState.totalBorrow.toString()).div(decimalScaleFactorBorrowToken);
    const vaultCollateral = new BN(vaultState.totalSupply.toString()).div(decimalScaleFactorSupplyToken);

    expect(topTick).to.be.eq(topmostTick.toString());
    expect(vaultCollateral.toString()).to.be.eq(totalCollateral.toString());
    expect(vaultDebt.toString()).to.be.eq(totalDebt.toString());

    const tickData = await setup.vaultModule.readTick({ vaultId, tick });
    const debtInTick = tickData.rawDebt.div(decimalScaleFactorBorrowToken);
    expect(debtInTick.toString()).to.be.eq(tickDebt.toString());
  }

  it("should handle deposit + borrow in single transaction", async () => {
    const vaultId = 1;

    const collateralAmount = new BN("1000000000"); // deposit 1000 USDC
    const debtAmount = new BN("500000000"); // borrow 500 USDT

    // prettier-ignore
    await createCheckedPosition(vaultId, collateralAmount, debtAmount, setup.alice, setup);

    // Verify vault state
    await _assertState(
      vaultId,
      -462, // Expected tick for the ratio
      new BN("500333316"), // Expected debt in tick (with some margin)
      -462,
      new BN("1000000000"), // Total collateral
      new BN("500000000") // Total debt
    );
  });

  it("should handle withdraw + payback in single transaction", async () => {
    const vaultId = 1;

    let collateral = new BN("1000000000");
    let debt = new BN("500000000");

    // prettier-ignore
    const positionId = await createCheckedPosition(vaultId, collateral, debt, setup.alice, setup);

    // as we are rounding up now, collateral amount becomes 1 more than the original collateral amount
    // so collateral amount that user transfers to the vault is 1000000001
    // but the amount stored in the position is 1000000000

    const withdrawAmount = new BN("300000000"); // withdraw 300 USDC
    const paybackAmount = new BN("200000000"); // payback 200 USDT

    // prettier-ignore
    // Perform withdraw + payback in single transaction
    await createCheckedPosition(vaultId, withdrawAmount.neg(), paybackAmount.neg(), setup.alice, setup, positionId);

    // Verify remaining position has 700 USDC collateral and 300 USDT debt
    await _assertState(
      vaultId,
      -565, // New tick after partial payback
      new BN("300129883"), // Expected remaining debt in tick
      -565,
      new BN("700000000"), // Remaining collateral
      new BN("300000000") // Remaining debt
    );
  });

  it("should handle deposit + payback in single transaction", async () => {
    const vaultId = 1;

    const collateralAmount = new BN("1000000000");
    const debtAmount = new BN("500000000");

    // First create a position with collateral and debt
    // prettier-ignore
    const positionId = await createCheckedPosition(vaultId, collateralAmount, debtAmount, setup.alice, setup);

    const additionalCollateral = new BN("500000000"); // deposit 500 more USDC
    const paybackAmount = new BN("200000000"); // payback 200 USDT

    const totalCollateralFromThis =
      await setup.createPositionInEveryTickArrayRange(
        vaultId,
        DEFAULT_ORACLE_PRICE
      );

    // prettier-ignore
    await createCheckedPosition(vaultId, additionalCollateral, paybackAmount.neg(), setup.alice, setup, positionId);

    // Verify position now has 1500 USDC collateral and 300 USDT debt
    await _assertState(
      vaultId,
      -1073, // New tick after adding collateral and reducing debt
      new BN("300343345"), // Expected remaining debt in tick
      -1073,
      new BN("1500000000").add(totalCollateralFromThis), // Total collateral after deposit
      new BN("49142585008") // Remaining debt after payback
    );
  });

  it("should handle withdraw + borrow in single transaction (if valid)", async () => {
    const vaultId = 1;

    let supplyAmount = new BN("2000000000");
    let debtAmount = new BN(0);

    const positionId = await createCheckedPosition(
      vaultId,
      supplyAmount,
      debtAmount,
      setup.alice,
      setup
    );

    // as we are rounding up now, supply amount becomes 1 more than the original supply amount
    // so supply amount is 2000000001

    const withdrawAmount = new BN("500000000").neg(); // withdraw 500 USDC
    const borrowAmount = new BN("400000000"); // borrow 400 USDT

    // prettier-ignore
    await createCheckedPosition(vaultId, withdrawAmount, borrowAmount, setup.alice, setup, positionId);

    // Verify position now has 1500 USDC collateral and 400 USDT debt
    await _assertState(
      vaultId,
      -881, // Expected tick for the new ratio
      new BN("400498700"), // Expected debt in tick
      -881,
      new BN("1500000000"), // Remaining collateral after withdrawal, original collateral rounded up during tranfers
      new BN("400000000") // Total debt after borrow
    );
  });

  it("should reject withdraw + borrow if it makes position unsafe", async () => {
    const vaultId = 1;

    let supplyAmount = new BN("1000000000");
    let debtAmount = new BN(0);

    // prettier-ignore
    const positionId = await createCheckedPosition(vaultId, supplyAmount, debtAmount, setup.alice, setup);

    // Try to withdraw too much while borrowing (should exceed collateral factor)
    const withdrawAmount = new BN("800000000"); // withdraw 800 USDC
    const borrowAmount = new BN("700000000"); // borrow 700 USDT

    expect(
      await setup.expectRevert(
        "VAULT_POSITION_ABOVE_CF",
        setup.operateVault.bind(setup),
        {
          vaultId,
          positionId,
          user: setup.alice,
          positionOwner: setup.alice,
          collateralAmount: withdrawAmount.neg(),
          debtAmount: borrowAmount,
          recipient: setup.alice,
        }
      )
    ).to.be.true;
  });

  it("should allow non-owner to deposit + payback in single transaction", async () => {
    const vaultId = 1;

    let collateral = new BN("1000000000");
    let debt = new BN("500000000");

    // prettier-ignore
    const positionId = await createCheckedPosition(vaultId, collateral, debt, setup.alice, setup);

    let additionalCollateral = new BN("300000000"); // Bob deposits 300 USDC
    let paybackAmount = new BN("200000000"); // Bob pays back 200 USDT

    const totalCollateralFromThis =
      await setup.createPositionInEveryTickArrayRange(
        vaultId,
        DEFAULT_ORACLE_PRICE
      );

    const bobSupplyTokenBefore = await setup.balanceOf(
      setup.bob.publicKey,
      MintInfo.getMint(setup.getVaultSupplyToken(vaultId))
    );

    const bobBorrowTokenBefore = await setup.balanceOf(
      setup.bob.publicKey,
      MintInfo.getMint(setup.getVaultBorrowToken(vaultId))
    );

    // Bob can deposit + payback on Alice's position
    await setup.operateVault({
      vaultId,
      positionId,
      user: setup.bob,
      positionOwner: setup.alice,
      collateralAmount: additionalCollateral,
      debtAmount: paybackAmount.neg(),
      recipient: setup.bob,
    });

    const bobSupplyTokenAfter = await setup.balanceOf(
      setup.bob.publicKey,
      MintInfo.getMint(setup.getVaultSupplyToken(vaultId))
    );

    const bobBorrowTokenAfter = await setup.balanceOf(
      setup.bob.publicKey,
      MintInfo.getMint(setup.getVaultBorrowToken(vaultId))
    );

    // round up the additional collateral during deposit
    if (additionalCollateral.gt(new BN(0))) {
      additionalCollateral = additionalCollateral.add(new BN(1));
    }

    // round down the payback amount during payback
    if (paybackAmount.gt(new BN(0))) {
      // +1 to the negative payback amount becomes extra payback
      paybackAmount = paybackAmount.add(new BN(1));
    }

    // Check Bob's token balances changed correctly
    expect(bobSupplyTokenBefore.sub(bobSupplyTokenAfter).toString()).to.be.eq(
      additionalCollateral.toString()
    );

    expect(bobBorrowTokenBefore.sub(bobBorrowTokenAfter).toString()).to.be.eq(
      paybackAmount.toString()
    );

    // Verify position state improved
    await _assertState(
      vaultId,
      -978, // New safer tick
      new BN("300130894"), // Remaining debt
      -978,
      new BN("1300000000").add(totalCollateralFromThis), // Total collateral
      new BN("49142585008") // Remaining debt
    );
  });

  it("should reject non-owner trying withdraw + borrow", async () => {
    const vaultId = 1;

    let collateral = new BN("2000000000");
    let debt = new BN("0");

    // Alice creates a position
    // prettier-ignore
    const positionId = await createCheckedPosition(vaultId, collateral, debt, setup.alice, setup);

    // Bob tries to withdraw + borrow from Alice's position (should fail)
    expect(
      await setup.expectRevert(
        "VAULT_INVALID_POSITION_AUTHORITY",
        setup.operateVault.bind(setup),
        {
          vaultId,
          positionId,
          user: setup.bob,
          positionOwner: setup.alice,
          collateralAmount: new BN("500000000").neg(), // withdraw
          debtAmount: new BN("300000000"), // borrow
          recipient: setup.bob,
        }
      )
    ).to.be.true;
  });
});

File: programs/fluid-solana-programs/tests/src/vaults/combination_test.rs
//! Vault Combination Tests - Rust port of TypeScript `combination.test.ts`.
//!
//! This module contains tests for vault operations that combine multiple
//! operations like deposit+borrow, withdraw+payback in single transactions.

#[cfg(test)]
mod tests {
    use crate::vaults::fixture::{
        OperateVars, VaultFixture, DEFAULT_ORACLE_PRICE,
    };
    use solana_sdk::signer::Signer;

    fn setup_vault_fixture() -> VaultFixture {
        let mut fixture = VaultFixture::new().expect("Failed to create vault fixture");
        fixture.setup().expect("Failed to setup vault fixture");
        // Set oracle price to be 1e8 for USDC/USDT (1:1 ratio)
        fixture
            .set_oracle_price(DEFAULT_ORACLE_PRICE, true)
            .expect("Failed to set oracle price");
        fixture
    }

    fn create_checked_position(
        fixture: &mut VaultFixture,
        vault_id: u16,
        collateral: i128,
        debt: i128,
        user: &solana_sdk::signature::Keypair,
        position_id: Option<u32>,
    ) -> u32 {
        let pos_id = if let Some(id) = position_id {
            id
        } else {
            let next_id = fixture
                .get_next_position_id(vault_id)
                .expect("Failed to get next position id");
            fixture
                .init_position(vault_id, user)
                .expect("Failed to init position");
            next_id
        };

        let supply_mint = fixture.get_vault_supply_token(vault_id);
        let borrow_mint = fixture.get_vault_borrow_token(vault_id);

        let user_supply_balance_before = fixture.balance_of(&user.pubkey(), supply_mint);
        let user_borrow_balance_before = fixture.balance_of(&user.pubkey(), borrow_mint);

        fixture
            .operate_vault(&OperateVars {
                vault_id,
                position_id: pos_id,
                user,
                position_owner: user,
                collateral_amount: collateral,
                debt_amount: debt,
                recipient: user,
            })
            .expect("Failed to operate vault");

        let user_supply_balance_after = fixture.balance_of(&user.pubkey(), supply_mint);
        let user_borrow_balance_after = fixture.balance_of(&user.pubkey(), borrow_mint);

        // Verify balance changes
        if collateral > 0 {
            assert_eq!(
                user_supply_balance_before - user_supply_balance_after,
                collateral as u64 + 1, // Allow for 1 lamport difference due to rounding/rent
                "Supply balance difference should equal collateral amount"
            );
        } else if collateral < 0 {
            assert_eq!(
                user_supply_balance_after - user_supply_balance_before,
                collateral.unsigned_abs() as u64,
                "Supply balance increase should equal withdraw amount"
            );
        }

        if debt > 0 {
            assert_eq!(
                user_borrow_balance_after - user_borrow_balance_before,
                debt as u64,
                "Borrow balance increase should equal debt amount"
            );
        } else if debt < 0 {
            assert_eq!(
                user_borrow_balance_before - user_borrow_balance_after,
                debt.unsigned_abs() as u64 + 1, // Allow for 1 lamport difference due to rounding/rent
                "Borrow balance decrease should equal payback amount"
            );
        }

        pos_id
    }

    /// Test: should handle deposit + borrow in single transaction
    #[test]
    fn test_should_handle_deposit_and_borrow_in_single_transaction() {
        let mut fixture = setup_vault_fixture();
        let vault_id = 1u16;
        let alice = fixture.liquidity.alice.insecure_clone();

        let collateral_amount: i128 = 1_000_000_000; // deposit 1000 USDC
        let debt_amount: i128 = 500_000_000; // borrow 500 USDT

        create_checked_position(&mut fixture, vault_id, collateral_amount, debt_amount, &alice, None);

        // Verify vault state
        fixture.assert_state(
            vault_id,
            -462,          // Expected tick for the ratio
            500_333_316,   // Expected debt in tick (with some margin)
            -462,          // topmost tick
            1_000_000_000, // Total collateral
            500_000_000,   // Total debt
        ).expect("Failed to assert state");
    }

    /// Test: should handle withdraw + payback in single transaction
    #[test]
    fn test_should_handle_withdraw_and_payback_in_single_transaction() {
        let mut fixture = setup_vault_fixture();
        let vault_id = 1u16;
        let alice = fixture.liquidity.alice.insecure_clone();

        let collateral: i128 = 1_000_000_000;
        let debt: i128 = 500_000_000;

        let position_id =
            create_checked_position(&mut fixture, vault_id, collateral, debt, &alice, None);

        let withdraw_amount: i128 = -300_000_000; // withdraw 300 USDC
        let payback_amount: i128 = -200_000_000; // payback 200 USDT

        // Perform withdraw + payback in single transaction
        create_checked_position(
            &mut fixture,
            vault_id,
            withdraw_amount,
            payback_amount,
            &alice,
            Some(position_id),
        );

        // Verify remaining position has 700 USDC collateral and 300 USDT debt
        fixture.assert_state(
            vault_id,
            -565,
            300_129_883,
            -565,
            700_000_000,
            300_000_000,
        ).expect("Failed to assert state");
    }

    /// Test: should handle deposit + payback in single transaction
    #[test]
    fn test_should_handle_deposit_and_payback_in_single_transaction() {
        let mut fixture = setup_vault_fixture();
        let vault_id = 1u16;
        let alice = fixture.liquidity.alice.insecure_clone();

        let collateral_amount: i128 = 1_000_000_000;
        let debt_amount: i128 = 500_000_000;

        // First create a position with collateral and debt
        let position_id = create_checked_position(
            &mut fixture,
            vault_id,
            collateral_amount,
            debt_amount,
            &alice,
            None,
        );

        let additional_collateral: i128 = 500_000_000; // deposit 500 more USDC
        let payback_amount: i128 = 200_000_000; // payback 200 USDT

        let supply_scale_factor =
            fixture.get_decimal_scale_factor(fixture.get_vault_supply_token(vault_id));
        let borrow_scale_factor =
            fixture.get_decimal_scale_factor(fixture.get_vault_borrow_token(vault_id));
        fixture
            .create_position_in_every_tick_array_range(vault_id, DEFAULT_ORACLE_PRICE)
            .expect("Failed to create positions in every tick array range");
        let state_after_helper = fixture.read_vault_state(vault_id).unwrap();
        let collateral_after_helper = state_after_helper.total_supply / supply_scale_factor;
        let debt_after_helper = state_after_helper.total_borrow / borrow_scale_factor;

        create_checked_position(
            &mut fixture,
            vault_id,
            additional_collateral,
            -payback_amount,
            &alice,
            Some(position_id),
        );

        let expected_total_collateral =
            collateral_after_helper + additional_collateral as u128;
        let expected_total_debt = debt_after_helper - payback_amount as u128;
        fixture
            .assert_state(
                vault_id,
                -1073,
                300_343_345,
                -1073,
                expected_total_collateral,
                expected_total_debt,
            )
            .expect("Failed to assert state");
    }

    /// Test: should handle withdraw + borrow in single transaction (if valid)
    #[test]
    fn test_should_handle_withdraw_and_borrow_in_single_transaction() {
        let mut fixture = setup_vault_fixture();
        let vault_id = 1u16;
        let alice = fixture.liquidity.alice.insecure_clone();

        let supply_amount: i128 = 2_000_000_000;
        let debt_amount: i128 = 0;

        let position_id =
            create_checked_position(&mut fixture, vault_id, supply_amount, debt_amount, &alice, None);

        let withdraw_amount: i128 = -500_000_000; // withdraw 500 USDC
        let borrow_amount: i128 = 400_000_000; // borrow 400 USDT

        create_checked_position(
            &mut fixture,
            vault_id,
            withdraw_amount,
            borrow_amount,
            &alice,
            Some(position_id),
        );

        // Verify position now has 1500 USDC collateral and 400 USDT debt
        fixture.assert_state(
            vault_id,
            -881,
            400_498_700,
            -881,
            1_500_000_000,
            400_000_000,
        ).expect("Failed to assert state");
    }

    /// Test: should reject withdraw + borrow if it makes position unsafe
    #[test]
    fn test_should_reject_withdraw_borrow_if_unsafe() {
        let mut fixture = setup_vault_fixture();
        let vault_id = 1u16;
        let alice = fixture.liquidity.alice.insecure_clone();

        let supply_amount: i128 = 1_000_000_000;
        let debt_amount: i128 = 0;

        let position_id =
            create_checked_position(&mut fixture, vault_id, supply_amount, debt_amount, &alice, None);

        // Try to withdraw too much while borrowing (should exceed collateral factor)
        let withdraw_amount: i128 = -800_000_000; // withdraw 800 USDC
        let borrow_amount: i128 = 700_000_000; // borrow 700 USDT

        let result = fixture.operate_vault(&OperateVars {
            vault_id,
            position_id,
            user: &alice,
            position_owner: &alice,
            collateral_amount: withdraw_amount,
            debt_amount: borrow_amount,
            recipient: &alice,
        });

        assert!(result.is_err(), "Should fail: VAULT_POSITION_ABOVE_CF");
        let err_str = format!("{:?}", result.unwrap_err());
        assert!(
            err_str.contains("VAULT_POSITION_ABOVE_CF") || err_str.contains("PositionAboveCF"),
            "Error should indicate position above collateral factor"
        );
    }

    /// Test: should allow non-owner to deposit + payback in single transaction
    #[test]
    fn test_should_allow_non_owner_deposit_and_payback() {
        let mut fixture = setup_vault_fixture();
        let vault_id = 1u16;
        let alice = fixture.liquidity.alice.insecure_clone();
        let bob = fixture.liquidity.bob.insecure_clone();

        let collateral: i128 = 1_000_000_000;
        let debt: i128 = 500_000_000;

        let position_id =
            create_checked_position(&mut fixture, vault_id, collateral, debt, &alice, None);

        let additional_collateral: i128 = 300_000_000; // Bob deposits 300 USDC
        let payback_amount: i128 = 200_000_000; // Bob pays back 200 USDT

        let supply_mint = fixture.get_vault_supply_token(vault_id);
        let borrow_mint = fixture.get_vault_borrow_token(vault_id);
        let supply_scale_factor = fixture.get_decimal_scale_factor(supply_mint);
        let borrow_scale_factor = fixture.get_decimal_scale_factor(borrow_mint);
        fixture
            .create_position_in_every_tick_array_range(vault_id, DEFAULT_ORACLE_PRICE)
            .expect("Failed to create positions in every tick array range");
        let state_after_helper = fixture.read_vault_state(vault_id).unwrap();
        let collateral_after_helper = state_after_helper.total_supply / supply_scale_factor;
        let debt_after_helper = state_after_helper.total_borrow / borrow_scale_factor;
        let expected_total_collateral =
            collateral_after_helper + additional_collateral as u128;

        let bob_supply_before = fixture.balance_of(&bob.pubkey(), supply_mint);
        let bob_borrow_before = fixture.balance_of(&bob.pubkey(), borrow_mint);

        // Bob can deposit + payback on Alice's position
        fixture
            .operate_vault(&OperateVars {
                vault_id,
                position_id,
                user: &bob,
                position_owner: &alice,
                collateral_amount: additional_collateral,
                debt_amount: -payback_amount,
                recipient: &bob,
            })
            .expect("Bob should be able to deposit + payback on Alice's position");

        let bob_supply_after = fixture.balance_of(&bob.pubkey(), supply_mint);
        let bob_borrow_after = fixture.balance_of(&bob.pubkey(), borrow_mint);

        // Check Bob's token balances changed correctly
        assert_eq!(
            bob_supply_before - bob_supply_after,
            additional_collateral as u64 + 1, // Allow for 1 lamport difference due to rounding/rent
            "Bob's supply balance should decrease by deposit amount"
        );
        assert_eq!(
            bob_borrow_before - bob_borrow_after,
            payback_amount as u64 + 1, // Allow for 1 lamport difference due to rounding/rent
            "Bob's borrow balance should decrease by payback amount"
        );
        let expected_total_debt = debt_after_helper - payback_amount as u128;
        fixture
            .assert_state(
                vault_id,
                -978,
                300_130_894,
                -978,
                expected_total_collateral,
                expected_total_debt,
            )
            .expect("Failed to assert state");
    }

    /// Test: should reject non-owner trying withdraw + borrow
    #[test]
    fn test_should_reject_non_owner_withdraw_borrow() {
        let mut fixture = setup_vault_fixture();
        let vault_id = 1u16;
        let alice = fixture.liquidity.alice.insecure_clone();
        let bob = fixture.liquidity.bob.insecure_clone();

        let collateral: i128 = 2_000_000_000;
        let debt: i128 = 0;

        // Alice creates a position
        let position_id =
            create_checked_position(&mut fixture, vault_id, collateral, debt, &alice, None);

        // Bob tries to withdraw + borrow from Alice's position (should fail)
        let result = fixture.operate_vault(&OperateVars {
            vault_id,
            position_id,
            user: &bob,
            position_owner: &alice,
            collateral_amount: -500_000_000, // withdraw
            debt_amount: 300_000_000,        // borrow
            recipient: &bob,
        });

        assert!(result.is_err(), "Should fail: VAULT_INVALID_POSITION_AUTHORITY");
        let err_str = format!("{:?}", result.unwrap_err());
        assert!(
            err_str.contains("VAULT_INVALID_POSITION_AUTHORITY")
                || err_str.contains("InvalidPositionAuthority"),
            "Error should indicate invalid position authority"
        );
    }
}

